go/vt/vtgate/endtoend/vstream_test.go

```diff
@@ -210,19 +210,39 @@ func TestVStreamCopyBasic(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	numExpectedEvents := 2 /* num shards */ * (7 /* begin/field/vgtid:pos/2 rowevents avg/vgitd: lastpk/commit) */ + 3 /* begin/vgtid/commit for completed table */)
+	numExpectedEvents := 2 /* num shards */ *(7 /* begin/field/vgtid:pos/2 rowevents avg/vgitd: lastpk/commit) */ +3 /* begin/vgtid/commit for completed table */ +1 /* copy operation completed */) + 1 /* fully copy operation completed */
+	expectedCompletedEvents := []string{
+		`type:COPY_COMPLETED keyspace:"ks" shard:"80-"`,
+		`type:COPY_COMPLETED keyspace:"ks" shard:"-80"`,
+		`type:COPY_COMPLETED`,
+	}
 	require.NotNil(t, reader)
 	var evs []*binlogdatapb.VEvent
+	var completedEvs []*binlogdatapb.VEvent
 	for {
 		e, err := reader.Recv()
 		switch err {
 		case nil:
 			evs = append(evs, e...)
+
+			for _, ev := range e {
+				if ev.Type == binlogdatapb.VEventType_COPY_COMPLETED {
+					completedEvs = append(completedEvs, ev)
+				}
+			}
+
+			printEvents(evs) // for debugging ci failures
+
 			if len(evs) == numExpectedEvents {
+				sort.Sort(VEventShardSorter(completedEvs))
+				for i, ev := range completedEvs {
+					require.Regexp(t, expectedCompletedEvents[i], ev.String())
+				}
 				t.Logf("TestVStreamCopyBasic was successful")
 				return
+			} else if numExpectedEvents < len(evs) {
+				t.Fatalf("len(events)=%v are not expected\n", len(evs))
```

Comment:
Previously, once `len(evs)` gets greater than `numExpectedEvents,` `End-to-End Test` job failed due to the timeout while showing no specific errors. With this error message, it's hard to narrow down the cause.

```
ok  	vitess.io/vitess/go/vt/vtctl/grpcvtctldserver/endtoend	0.130s
Error: The action has timed out.
```

- ref. https://github.com/vitessio/vitess/actions/runs/3477398228/jobs/5814821217

[5825b27](https://github.com/vitessio/vitess/pull/11740/commits/5825b2738a05c5b3519718168859171219fa5b4c) makes the same test fail faster and shows more straightforward error messages.

```
--- FAIL: TestVStreamCopyBasic (0.02s)
    vstream_test.go:244: len(events)=24 are not expected
...
FAIL	vitess.io/vitess/go/vt/vtgate/endtoend	77.838s
```

- ref. https://github.com/vitessio/vitess/actions/runs/3485888371/jobs/5832080670